### PR TITLE
En Passant

### PR DESCRIPTION
**Pull request by takehiromouri**
# En Passant
* [Overview of En Passant](http://www.chess.com/chessopedia/view/en-passant)
  * En passant (from French: "in passing") is a maneuver in chess which is performed after a player moves a pawn two squares forward from its starting position, and an opposing pawn captures it as if it had only moved one square.  En passant may only be played immediately after a two-square square pawn advance, or the right to capture "in passing" is lost.
* Checks to see if En Passant is possible in `valid_move?`
  * Checks if there are adjacent Pawns that match the last updated Piece
  * Checks if that Pawn moved 2 squares in the previous move
* Checks to see if En Passant is possible in `move_to!`
  * If it is possible, capture!

### Testing
* Valid En Passant Move
* En Passant attempt when the last updated Pawn only moved 1 square in previous turn
* En Passant moves when destination coordinate is obstructed (should still capture piece)
* En Passant attempt when the adjacent piece is not a pawn
* En Passant move when there are two adjacent pieces

### CODE REVIEW

File: app/models/piece.rb [hunk 1]
```diff
@@ -13,6 +14,22 @@ def move_to!(new_row, new_col)
     end
   end
 
+  def capture_en_passant(row_dest, col_dest)
+    @last_updated = Piece.where(game_id: game_id).order("updated_at desc").first
+    return if @last_updated.nil?
```
Comment (jonathanpike) [inline app/models/piece.rb]: You should probably return here if `@last_updated` is not a Pawn (ie. `return if @last_updated.nil? || @last_updated.type != "Pawn"`)

File: app/models/piece.rb [hunk 2]
```diff
@@ -13,6 +14,22 @@ def move_to!(new_row, new_col)
...
+    if Game.find(game_id).black_player_id == user_id 
+       puts "Hi"
+      if row_dest == (@last_updated.row_position - 1) && col_dest == @last_updated.col_position
```
Comment (jonathanpike) [inline app/models/piece.rb]: Should you not first check if your piece is beside `@last_updated`?  That would let you know if you could capture or not.

File: app/models/piece.rb (hunk 1 above)
Comment (takehiromouri) [inline app/models/piece.rb]: This is where I get confused because I already check this in valid_move? so I am wondering if this is necessary

**Comment by jonathanpike** [comment]
I can totally see why this is confusing, @takehiromouri. I find it helpful to break things down into the simplest possible expression of the problem, and maybe you will too. Here are the steps that have to happen for an en passant capture, as far as I understand them: 

1. Check if your pawn is beside another pawn that just moved 2 rows in the last move
2. If so, capture! 

Things we have to check: 

1. Opposing pawn is beside you (easy -- just compare `row_position` and `col_position`)
2. Opposing pawn just moved (easy -- just check if it was the piece that was last updated in the DB)
3. Opposing pawn just moved 2 rows (harder)

The first 2 could be implemented like you already have: 

```
@last_updated = Piece.where(game_id: game_id).order("updated_at desc").first
return unless @last_updated.type == "Pawn" && @last_updated.row_position == row_position && ( @last_updated.col_position == (col_position + 1) || @last_updated.col_position == (col_position - 1)
```

You've implemented a `previous_changes_hash` that looks like it should take care of 3.  However, I'd suggest looking into [ActiveModel::Dirty](http://api.rubyonrails.org/classes/ActiveModel/Dirty.html), which seems to do this without having to implement it as a new column in our DB. 

If we use that, you could just call `@last_updated.changes` and see if it moved 2 rows in the last turn. 

So, from there, what do we need to implement? 

1. valid_move? for Pawn.  If those 3 checks pass, allow a diagonal move to an empty square.
2. move_to!  This will need to include the same 3 checks and change `@last_updated` to captured. 

I hope this helps a little.  I'd love to be able to pair with you if you want to work together with someone.

**Pull request merged**